### PR TITLE
Use locks for mutable state within JobManager

### PR DESCRIPTION
**Pull request by wcharczuk**
This PR creates locks for each map we write to within JobManager. Deadlocks inbound.

### CODE REVIEW

File: job_manager.go [hunk 1]
```diff
@@ -65,78 +62,305 @@ func Default() *JobManager {
...
+	if !jm.HasJob(jobName) {
 		return exception.Newf("Job name `%s` isn't loaded.", jobName)
 	}
 
-	jm.metaLock.Lock()
-	defer jm.metaLock.Unlock()
-
-	jm.DisabledJobs.Add(jobName)
-	delete(jm.NextRunTimes, jobName)
+	jm.SetDisabledJob(jobName)
```
Comment (christopheralme) [inline job_manager.go]: Does this need to have jm.HasJob be true? I'm pretty sure the answer is no, just pointing out that as written, it this line could technically get undercut.

**Comment by christopheralme** [comment]
lgtm. The main thing I'm concerned about is deadlocking when multiple lock are acquired. Is there any deadlock detection stuff that exists? https://golang.org/doc/articles/race_detector.html? Make -race part of make test if it isn't already?

File: job_manager.go [hunk 2]
```diff
@@ -65,78 +62,305 @@ func Default() *JobManager {
...
+
+// DeleteRunningTask deletes jm.runningTasks
+func (jm *JobManager) DeleteRunningTask(taskName string) {
+	jm.runningTasksLock.Lock()
+	defer jm.runningTasksLock.Unlock()
+
+	delete(jm.runningTasks, taskName)
+}
+
+// GetRunningTaskStartTime gets jm.runningTaskStartTimess
```
Comment (christopheralme) [inline job_manager.go]: Not used?

File: job_manager.go [hunk 3]
```diff
@@ -65,78 +62,305 @@ func Default() *JobManager {
...
+
+// SetLoadedJob sets jm.loadedJobs
+func (jm *JobManager) SetLoadedJob(jobName string, j Job) {
+	jm.loadedJobsLock.Lock()
+	defer jm.loadedJobsLock.Unlock()
+
+	jm.loadedJobs[jobName] = j
+}
+
+// DeleteLoadedJob deletes jm.loadedJobs
```
Comment (christopheralme) [inline job_manager.go]: not used?

File: job_manager.go [hunk 4]
```diff
@@ -65,78 +62,305 @@ func Default() *JobManager {
...
+// DeleteLastRunTime deletes jm.LastRunTimes
+func (jm *JobManager) DeleteLastRunTime(taskName string) {
+	jm.lastRunTimesLock.Lock()
+	defer jm.lastRunTimesLock.Unlock()
+
+	delete(jm.lastRunTimes, taskName)
+}
+
+// GetLoadedJob gets a job
+func (jm *JobManager) GetLoadedJob(jobName string) Job {
```
Comment (christopheralme) [inline job_manager.go]: Not used

File: job_manager.go (hunk 2 above)
Comment (wcharczuk) [inline job_manager.go]: not used by this package, my fear is descendant packages might need this.

**Comment by wcharczuk** [comment]
The key is we need to simulate deadlock situation in tests. I'm going to take some time today to simulate deadlocking situations in tests.

File: job_manager.go [hunk 5]
```diff
@@ -286,57 +528,107 @@ func (jm *JobManager) Stop() {
...
-					if now.Sub(startedTime) >= timeout {
-						jm.CancelTask(taskName)
-						if receiver, isReceiver := task.(OnCompleteReceiver); isReceiver {
-							receiver.OnComplete(exception.New("Timeout Reached."))
-						}
-					}
+func (jm *JobManager) killHangingJobsInner() {
+    
+    jm.runningTasksLock.Lock()
+	defer jm.runningTasksLock.Unlock()
```
Comment (christopheralme) [inline job_manager.go]: weird tabbing

File: job_manager_test.go [hunk 6]
```diff
@@ -2,7 +2,9 @@ package chronometer
 
 import (
 	"testing"
+    "fmt"
```
Comment (christopheralme) [inline job_manager_test.go]: not aligned

File: job_manager.go [hunk 7]
```diff
@@ -286,57 +528,107 @@ func (jm *JobManager) Stop() {
...
 				}
 			}
 		}
-		time.Sleep(HeartbeatInterval)
 	}
 }
 
+// killHangingJob cancels (sends the cancellation signal) to a running task that has exceeded its timeout.
+func (jm *JobManager) killHangingJob(taskName string) error {
+	if task, hasTask := jm.runningTasks[taskName]; hasTask {
```
Comment (christopheralme) [inline job_manager.go]: if we're locking on all reads, we should add a comment about the caller needing to have acquired the runningTasksLock and cancellationTokensLock to call this method, or explicitly move it into this method.

File: job_manager.go [hunk 8]
```diff
@@ -65,78 +65,305 @@ func Default() *JobManager {
...
-
-	job, hasJob := jm.LoadedJobs[jobName]
-	if !hasJob {
+	if !jm.HasJob(jobName) {
 		return exception.Newf("Job name `%s` isn't loaded.", jobName)
 	}
+
+	jm.DeleteDisabledJob(jobName)
+
+	job := jm.loadedJobs[jobName]
```
Comment (christopheralme) [inline job_manager.go]: read on loaded job without lock? Don't think codepath is used

Comment (wcharczuk) [inline job_manager.go]: Enable / Disable is used outside chronometer (by referencing packages etc.)

**Pull request merged**